cpp/src/arrow/telemetry/logging.cc

```diff
@@ -0,0 +1,294 @@
+#include <opentelemetry/proto/collector/logs/v1/logs_service.pb.h>
+
+#include <opentelemetry/exporters/otlp/protobuf_include_suffix.h>
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+namespace arrow {
+namespace telemetry {
+
+namespace {
+
+using internal::OtelLogExporterOptions;
+
+constexpr const char kLoggingBackendEnvVar[] = "ARROW_LOGGING_BACKEND";
+
+class OtlpOStreamLogRecordExporter final : public otel::sdk::logs::LogRecordExporter {
+ public:
+  explicit OtlpOStreamLogRecordExporter(std::ostream* sink) : sink_(sink) {
+    pb_json_options_.add_whitespace = false;
+  }
+
+  otel::sdk::common::ExportResult Export(
+      const otel_span<std::unique_ptr<otel::sdk::logs::Recordable>>& records) noexcept
+      override {
+    otel::proto::collector::logs::v1::ExportLogsServiceRequest request;
+    otel::exporter::otlp::OtlpRecordableUtils::PopulateRequest(records, &request);
+
+    for (const auto& logs : request.resource_logs()) {
+      std::string out;
+      auto status =
+          google::protobuf::util::MessageToJsonString(logs, &out, pb_json_options_);
+      if (ARROW_PREDICT_FALSE(!status.ok())) {
+        return otel::sdk::common::ExportResult::kFailure;
+      }
+      (*sink_) << out << std::endl;
+    }
+
+    return otel::sdk::common::ExportResult::kSuccess;
+  }
+
+  bool ForceFlush(std::chrono::microseconds timeout) noexcept override {
+    return exporter_.ForceFlush(timeout);
+  }
+
+  bool Shutdown(std::chrono::microseconds timeout) noexcept override {
+    return exporter_.Shutdown(timeout);
+  }
+
+  std::unique_ptr<otel::sdk::logs::Recordable> MakeRecordable() noexcept override {
+    return exporter_.MakeRecordable();
+  }
+
+ private:
+  std::ostream* sink_;
+  otel::exporter::otlp::OtlpHttpLogRecordExporter exporter_;
+  google::protobuf::util::JsonPrintOptions pb_json_options_;
+};
+
+otel::logs::Severity ToOtelSeverity(LogLevel level) {
+  switch (level) {
+    case LogLevel::ARROW_TRACE:
+      return opentelemetry::logs::Severity::kTrace;
+    case LogLevel::ARROW_DEBUG:
+      return opentelemetry::logs::Severity::kDebug;
+    case LogLevel::ARROW_INFO:
+      return opentelemetry::logs::Severity::kInfo;
+    case LogLevel::ARROW_WARNING:
+      return opentelemetry::logs::Severity::kWarn;
+    case LogLevel::ARROW_ERROR:
+      return opentelemetry::logs::Severity::kError;
+    case LogLevel::ARROW_FATAL:
+      return opentelemetry::logs::Severity::kFatal;
+  }
+  return otel::logs::Severity::kInvalid;
+}
+
+enum class ExporterKind {
+  OSTREAM,
+  OTLP_HTTP,
+  OTLP_OSTREAM,
+};
+
+std::unique_ptr<otel::sdk::logs::LogRecordExporter> MakeExporter(
+    ExporterKind exporter_kind, std::ostream* ostream = nullptr) {
+  switch (exporter_kind) {
+    case ExporterKind::OSTREAM: {
+      return std::make_unique<otel::exporter::logs::OStreamLogRecordExporter>(*ostream);
+    } break;
+    case ExporterKind::OTLP_HTTP: {
+      namespace otlp = otel::exporter::otlp;
+      // TODO: Allow user configuration here?
```

Comment:
Shouldn't this all be the end application's responsibility?

Reply:
Yep. Old comment.